src/crate_metadata.rs

```diff
@@ -41,11 +41,22 @@ pub struct CrateMetadata {
 
 impl CrateMetadata {
     /// Parses the contract manifest and returns relevant metadata.
-    pub fn collect(manifest_path: &ManifestPath) -> Result<Self> {
+    pub fn collect(
+        manifest_path: &ManifestPath,
+        build_artifact: Option<BuildArtifacts>,
```

Comment:
Rather than passing in the `BuildArtifact`, I think instead this should be `target_directory: Option<PathBuf>`, and then move the block with `CheckOnly` and passing `Some(tmp_dir)` into `cmd/build.rs`, otherwise `None`.

All other call sites will pass `None`. This moves the logic to earlier in the code path and also means that `CrateMetadata` doesn't need to know about `BuildArtifacts`

Reply:
Ok, so in general I agree about shifting that logic, but this function creates `CrateMetadata`, which needs to contain the path to `original_wasm` and `dest_wasm`, which both have to contain `target_directory`. So it should be `target_directory: PathBuf`, right?

Reply:
I think I meant that if you pass `Some(target_dir)` it uses that otherwise it still uses `metadata.target_directory`, something like:

```rust
let target_directory = target_directory.unwrap_or_else(|| 
    metadata.target_directory.clone().as_path().join("ink")
);
```